http/binding/binding_test.go

```diff
@@ -0,0 +1,472 @@
+	id, err := binding.BindID(c)
+	if err != nil {
+		c.AbortWithStatus(errors.Code(err))
+	}
+	c.JSON(http.StatusOK, id)
+}
+
+type BindingSuite struct {
+	suite.Suite
+	router *gin.Engine
+}
+
+// SetupTest runs before each Test
+func (s *BindingSuite) SetupTest() {
+	gin.SetMode(gin.TestMode)
+	s.router = gin.Default()
+}
+
+func TestHandlers(t *testing.T) {
+	suite.Run(t, new(BindingSuite))
+}
+
+func (s *BindingSuite) Test_BindJSON_FromBodyBindError() {
+	requestBody := `{
+	    "number": [
+	        0
+	    ],
+	    "string": [
+	        "🇺🇸",
+	    ]
+	}`
+	req, err := http.NewRequest(http.MethodPatch, testJSONEndpoint, strings.NewReader(requestBody))
+	s.NoError(err)
+
+	s.router.PATCH(testJSONEndpoint, testJSONHandler)
+	r := httptest.NewRecorder()
+	s.router.ServeHTTP(r, req)
+
+	s.Equal(http.StatusBadRequest, r.Code)
+}
+
+func (s *BindingSuite) Test_BindJSON_FromBody() {
+	requestBody := `{
+	    "number": [
+	        1,
+	        2,
+	        3
+	    ],
+	    "string": [
+	        "A",
+	        "B",
+	        "C"
+	    ]
+	}`
+	req, err := http.NewRequest(http.MethodPatch, testJSONEndpoint, strings.NewReader(requestBody))
+	s.NoError(err)
+
+	s.router.PATCH(testJSONEndpoint, testJSONHandler)
+	r := httptest.NewRecorder()
+	s.router.ServeHTTP(r, req)
+
+	s.Equal(http.StatusOK, r.Code)
+	s.Equal(r.Body.String(), "{\"number\":[1,2,3],\"string\":[\"A\",\"B\",\"C\"]}")
+}
+
+func (s *BindingSuite) Test_BindJSON_FromContext_OK() {
+	requestBody := `{
+	    "number": [
+	        1,
+	        2,
+	        3
+	    ],
+	    "string": [
+	        "A",
+	        "B",
+	        "C"
+	    ]
+	}`
+	req, err := http.NewRequest(http.MethodPatch, testJSONEndpoint, strings.NewReader(requestBody))
+	s.NoError(err)
+
+	testReq := &testStruct{
+		Number: []uint{1, 2, 3, 4},
+		String: []string{"A", "B", "C", "D"},
+	}
+	s.router.Use(func(c *gin.Context) {
+		c.Set(ctxKey, testReq)
+		c.Next()
+	})
+	s.router.PATCH(testJSONEndpoint, testJSONHandler)
+	r := httptest.NewRecorder()
+	s.router.ServeHTTP(r, req)
+
+	s.Equal(http.StatusOK, r.Code)
+
+	body, err := json.Marshal(&testReq)
+	s.NoError(err)
+	s.Equal(r.Body.String(), string(body))
+}
+
+func (s *BindingSuite) Test_BindJSON_FromContext_TypeMismatch() {
+	requestBody := `{
+	    "number": [
+	        1,
+	        2,
+	        3
+	    ],
+	    "string": [
+	        "A",
+	        "B",
+	        "C"
+	    ]
+	}`
+	testReq := &testStruct{
+		Number: []uint{1, 2, 3, 4},
+		String: []string{"A", "B", "C", "D"},
+	}
+
+	s.router.Use(func(c *gin.Context) {
+		c.Set(ctxKey, &testReq)
+		c.Next()
+	})
+
+	req, err := http.NewRequest(http.MethodPatch, testJSONEndpoint, strings.NewReader(requestBody))
+	s.NoError(err)
+
+	s.router.PATCH(testJSONEndpoint, testJSONHandler)
+	r := httptest.NewRecorder()
+	s.router.ServeHTTP(r, req)
+
+	s.Equal(http.StatusOK, r.Code)
+
+	s.Equal(r.Body.String(), "{\"number\":[1,2,3],\"string\":[\"A\",\"B\",\"C\"]}")
+}
+
+func (s *BindingSuite) Test_BindQuery_FromBodyBindError() {
```

Comment:
Shall we rename those test cases to `Test_BindQuery_FromURL` or something similar?
`FromBody` is not correct for these cases.